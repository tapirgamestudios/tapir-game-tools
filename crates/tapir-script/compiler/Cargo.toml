[package]
name = "compiler"
version = "0.1.0"
edition = "2024"

[dependencies]
agb_fixnum = "0.22"
ariadne = "0.5"
lalrpop-util = "0.22.2"
logos = "0.15"
num-traits = "0.2"
serde = { version = "1", features = ["derive"] }

bytecode = { path = "../bytecode" }
petgraph = "0.6.5"

[dev-dependencies]
insta = { version = "1.40", features = ["ron", "glob", "redactions"] }

[build-dependencies]
lalrpop = "0.22.2"

[profile.dev.package]
insta.opt-level = 3
similar.opt-level = 3
