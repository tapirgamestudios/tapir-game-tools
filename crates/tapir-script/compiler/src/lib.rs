#![deny(clippy::all)]
use std::path::Path;

use lalrpop_util::lalrpop_mod;

use reporting::Diagnostics;

mod ast;
mod compile;
mod lexer;
mod reporting;
mod tokens;
mod types;

lalrpop_mod!(grammar);

#[cfg(test)]
mod grammar_test;

pub use compile::{CompileSettings, Property};
pub use reporting::{format::DiagnosticCache, Message};
pub use types::Type;

pub fn compile(
    filename: impl AsRef<Path>,
    input: &str,
    compile_settings: CompileSettings,
) -> Result<CompileResult, Diagnostics> {
    let bytecode = compile::compile(filename, input, &compile_settings)?;

    let compiled = bytecode.data.into_boxed_slice();
    Ok(CompileResult {
        bytecode: compiled,
        event_handlers: bytecode.event_handlers.into_boxed_slice(),
        triggers: bytecode.triggers,
    })
}

pub struct CompileResult {
    pub bytecode: Box<[u32]>,
    pub event_handlers: Box<[EventHandler]>,
    pub triggers: Box<[Trigger]>,
}

pub struct EventHandler {
    pub name: String,
    pub bytecode_offset: usize,
    pub arguments: Box<[EventHandlerArgument]>,
}

pub struct Trigger {
    pub name: String,
    pub arguments: Box<[Type]>,
}

#[derive(Clone)]
pub struct EventHandlerArgument {
    pub name: String,
    pub ty: Type,
}
