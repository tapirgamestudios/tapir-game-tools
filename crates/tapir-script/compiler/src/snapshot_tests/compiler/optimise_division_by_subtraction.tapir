# optimise
fn divide(n: int, d: int) -> int {
    var r = n;
    var q = 0;

    loop {
        if r < d {
            return q;
        }

        r = r - d;
        q = q + 1;
    }
}

int_prop = divide(100, int_prop);