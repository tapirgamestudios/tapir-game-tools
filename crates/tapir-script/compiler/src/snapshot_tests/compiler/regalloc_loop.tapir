# Testing a register allocation bug.
#
# The register allocator will assign the same register
# to x, y and `true` resulting in the value in y to be
# overwritten with the number 1
var x = 3;

loop {
    var y = x;

    if true {}
}