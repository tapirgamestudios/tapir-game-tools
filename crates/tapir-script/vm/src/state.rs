use crate::TapirScript;

use agb_fixnum::Num;
use alloc::{boxed::Box, vec::Vec};

type Fix = Num<i32, 8>;

#[derive(Debug)]
pub(crate) struct State {
    pc: usize,
    stack: Vec<i32>,
    stack_offset: usize,
}

pub(crate) enum RunResult {
    Waiting,
    Finished,
    Spawn(Box<State>),
}

impl State {
    pub(crate) fn new(pc: usize, stack: Vec<i32>) -> Self {
        Self {
            pc,
            stack,
            stack_offset: 0,
        }
    }

    #[cfg(test)]
    pub(crate) fn stack(&self) -> &[i32] {
        &self.stack
    }

    pub(crate) fn run_until_wait(
        &mut self,
        bytecode: &[u32],
        properties: &mut dyn ObjectSafeProperties,
    ) -> RunResult {
        loop {
            let Some(&instr) = bytecode.get(self.pc) else {
                return RunResult::Finished;
            };

            let opcode = bytecode::opcode(instr).expect("Invalid bytecode instruction");

            self.pc += 1;

            use bytecode::Opcode as O;

            macro_rules! type1 {
                ($target:ident) => {
                    type1!($target, _a)
                };
                ($target:ident, $a:ident) => {
                    type1!($target, $a, _b)
                };
                ($target:ident, $a:ident, $b:ident) => {
                    let bytecode::Type1 {
                        target: $target,
                        a: $a,
                        b: $b,
                        ..
                    } = bytecode::Type1::decode(instr);
                };
            }

            macro_rules! binop {
                ($a:ident, $b:ident, $op:expr) => {{
                    type1!(target, a, b);
                    let $a = self.get_reg(a);
                    let $b = self.get_reg(b);
                    self.set_reg(target, $op);
                }};
            }

            match opcode {
                O::Mov => {
                    type1!(target, a);
                    self.set_reg(target, self.get_reg(a));
                }

                O::Add => binop!(a, b, a + b),
                O::Sub => binop!(a, b, a - b),
                O::Mul => binop!(a, b, a * b),
                O::RealMod => binop!(a, b, a.rem_euclid(b)),
                O::RealDiv => binop!(a, b, a / b), // FIXME: div_floor
                O::EqEq => binop!(a, b, (a == b).into()),
                O::NeEq => binop!(a, b, (a != b).into()),
                O::Gt => binop!(a, b, (a > b).into()),
                O::GtEq => binop!(a, b, (a >= b).into()),
                O::Lt => binop!(a, b, (a < b).into()),
                O::LtEq => binop!(a, b, (a <= b).into()),
                O::FixMul => binop!(a, b, (Fix::from_raw(a) * Fix::from_raw(b)).to_raw()),
                O::FixDiv => binop!(a, b, (Fix::from_raw(a) / Fix::from_raw(b)).to_raw()),

                O::GetProp => {
                    type1!(target, prop_index);
                    self.set_reg(target, properties.get_prop(prop_index));
                }
                O::SetProp => {
                    type1!(value, prop_index);
                    properties.set_prop(prop_index, self.get_reg(value));
                }

                O::Call => {
                    type1!(first_arg);
                    let reg_value = (first_arg as u32) << 24 | (self.pc as u32);
                    self.set_reg(first_arg, reg_value as i32);
                    self.stack_offset += first_arg as usize;
                }
                O::Spawn => {
                    type1!(first_arg, num_args);
                    let mut new_stack = Vec::with_capacity(num_args as usize + 1);
                    new_stack.push(-1);

                    let stack_to_copy_start = self.stack_offset + usize::from(first_arg + 1);
                    let stack_to_copy_end = stack_to_copy_start + usize::from(num_args);
                    new_stack.extend(&self.stack[stack_to_copy_start..stack_to_copy_end]);

                    let new_thread_pc = self.pc;
                    self.pc += 1; // skip over the jump instruction

                    return RunResult::Spawn(Box::new(Self::new(new_thread_pc, new_stack)));
                }
                O::Trigger => {
                    type1!(id, first_arg);

                    let stack_to_copy_start = self.stack_offset + usize::from(first_arg + 1);
                    properties.add_event(id, &self.stack[stack_to_copy_start..]);
                }

                O::JumpIf => {
                    type1!(test);

                    let test_value = self.get_reg(test);
                    if test_value == 0 {
                        self.pc += 1;
                    }
                }

                O::Ret => {
                    let value = self.get_reg(0);
                    if value == -1 {
                        return RunResult::Finished;
                    }

                    let value = value as u32;
                    let offset = value >> 24;
                    self.stack_offset -= offset as usize;

                    let new_pc = value & 0x00FF_FFFF;
                    self.pc = new_pc as usize + 1; // skip the jump instruction
                }
                O::Wait => {
                    return RunResult::Waiting;
                }

                O::LoadConstant => {
                    type1!(target);
                    let constant = bytecode[self.pc];
                    self.pc += 1;

                    self.set_reg(target, constant as i32);
                }

                O::Jump => {
                    let target = instr & 0x00FF_FFFF;
                    self.pc = target as usize;
                }
            }
        }
    }

    fn set_reg(&mut self, reg: u8, value: i32) {
        *self.reg(reg) = value;
    }

    fn get_reg(&self, reg: u8) -> i32 {
        let reg_id = self.reg_id(reg);
        self.stack.get(reg_id).copied().unwrap_or(0)
    }

    fn reg_id(&self, reg: u8) -> usize {
        self.stack_offset + reg as usize
    }

    fn reg(&mut self, reg: u8) -> &mut i32 {
        let reg_id = self.reg_id(reg);
        self.stack.resize(self.stack.len().max(reg_id + 1), 0);

        &mut self.stack[reg_id]
    }
}

pub(crate) trait ObjectSafeProperties {
    fn set_prop(&mut self, index: u8, value: i32);
    fn get_prop(&self, index: u8) -> i32;

    fn add_event(&mut self, index: u8, args: &[i32]);
}

pub(crate) struct ObjectSafePropertiesImpl<'a, T, U>
where
    T: TapirScript<EventType = U>,
{
    pub properties: &'a mut T,
    pub events: Vec<U>,
}

impl<'a, T, U> ObjectSafeProperties for ObjectSafePropertiesImpl<'a, T, U>
where
    T: TapirScript<EventType = U>,
{
    fn set_prop(&mut self, index: u8, value: i32) {
        self.properties.set_prop(index, value);
    }

    fn get_prop(&self, index: u8) -> i32 {
        self.properties.get_prop(index)
    }

    fn add_event(&mut self, index: u8, args: &[i32]) {
        self.events.push(self.properties.create_event(index, args));
    }
}
