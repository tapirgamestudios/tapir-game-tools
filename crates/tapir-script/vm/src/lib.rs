#![deny(clippy::all)]
#![no_std]
extern crate alloc;

mod state;

use alloc::{vec, vec::Vec};
use state::{ObjectSafeProperties, ObjectSafePropertiesImpl, State};

struct Vm<'a> {
    bytecode: &'a [u32],
    states: Vec<State>,
}

impl<'a> Vm<'a> {
    pub fn new(bytecode: &'a [u32]) -> Self {
        Self {
            bytecode,
            states: vec![State::new(0, vec![-1])],
        }
    }

    fn run_until_wait(&mut self, properties: &mut dyn ObjectSafeProperties) {
        let mut state_index = 0;
        while state_index < self.states.len() {
            match self.states[state_index].run_until_wait(self.bytecode, properties) {
                state::RunResult::Waiting => {
                    state_index += 1;
                }
                state::RunResult::Finished => {
                    self.states.swap_remove(state_index);
                }
                state::RunResult::Spawn(state) => {
                    self.states.push(*state);
                    // intentionally not increasing state_index to ensure that the spawning
                    // state continues to run.
                }
            }
        }
    }
}

/// # Safety
///
/// You should never implement this directly, and instead go through the derive macro
pub unsafe trait TapirScript {
    type EventType;

    fn script(self) -> Script<Self>
    where
        Self: Sized;

    fn set_prop(&mut self, index: u8, value: i32);
    fn get_prop(&self, index: u8) -> i32;

    fn create_event(&self, index: u8, args: &[i32]) -> Self::EventType;
}

pub struct Script<T: TapirScript> {
    vm: Vm<'static>,
    pub properties: T,
}

impl<T: TapirScript> Script<T> {
    pub fn new(properties: T, bytecode: &'static [u32]) -> Self {
        Self {
            vm: Vm::new(bytecode),
            properties,
        }
    }

    pub fn run(&mut self) -> Vec<T::EventType> {
        let mut object_safe_props = ObjectSafePropertiesImpl {
            properties: &mut self.properties,
            events: vec![],
        };

        self.vm.run_until_wait(&mut object_safe_props);

        object_safe_props.events
    }

    pub fn will_calling_run_do_anything(&self) -> bool {
        !self.vm.states.is_empty()
    }

    #[doc(hidden)]
    pub unsafe fn __private_trigger_event(&mut self, initial_stack: Vec<i32>, pc: usize) {
        self.vm.states.push(State::new(pc, initial_stack));
    }
}

#[cfg(test)]
mod test {
    extern crate std;

    use super::*;

    use std::fs;

    use alloc::string::ToString;
    use compiler::{CompileSettings, Property, Type};
    use insta::{assert_ron_snapshot, glob};
    use serde::Serialize;

    use crate::TapirScript;

    #[test]
    fn stack_snapshot_tests() {
        glob!("snapshot_tests", "stack/**/*.tapir", |path| {
            std::println!("{}", path.display());

            let input = fs::read_to_string(path).unwrap();

            let compiler_settings = CompileSettings {
                properties: vec![Property {
                    ty: Type::Int,
                    index: 0,
                    name: "int_prop".to_string(),
                }],
                enable_optimisations: true,
            };

            let bytecode = compiler::compile(path.file_name().unwrap(), &input, compiler_settings)
                .unwrap()
                .bytecode;

            let mut vm = Vm::new(&bytecode);
            let mut prop_object = PropObj { int_prop: 5 };

            let mut stack_at_waits = vec![];

            let mut max_iterations = 1000;

            while !vm.states.is_empty() && max_iterations >= 0 {
                let mut object_safe_props = ObjectSafePropertiesImpl {
                    properties: &mut prop_object,
                    events: vec![],
                };

                vm.run_until_wait(&mut object_safe_props);
                stack_at_waits.push((
                    vm.states
                        .iter()
                        .map(|state| state.stack().to_vec())
                        .collect::<Vec<_>>(),
                    prop_object.clone(),
                ));

                max_iterations -= 1;
            }

            if max_iterations == 0 {
                panic!("ran for over 1000 waits, something seems to have gone wrong...");
            }

            assert_ron_snapshot!(stack_at_waits);
        });
    }

    macro_rules! binop_test {
        ($($type: ident, $name:ident: ($code:tt, $expected:expr),)*) => {
            $(
                paste::paste! {
                    #[test]
                    fn [< binop_test_ $name >]() {
                        let compile_settings = CompileSettings {
                            properties: vec![Property {
                                ty: Type::$type,
                                index: 0,
                                name: "prop".to_string(),
                            }],
                            enable_optimisations: false,
                        };

                        let bytecode = compiler::compile(
                            concat!(stringify!($name), ".tapir"),
                            concat!("prop = ", $code, ";"),
                            compile_settings
                        ).unwrap().bytecode;

                        let mut vm = Vm::new(&bytecode);
                        let mut prop_object = PropObj {
                            int_prop: if Type::$type == Type::Int { 5 } else { 1 },
                        };

                        while !vm.states.is_empty() {
                            let mut object_safe_props = ObjectSafePropertiesImpl {
                                properties: &mut prop_object,
                                events: vec![],
                            };

                            vm.run_until_wait(&mut object_safe_props);
                        }

                        assert_eq!(prop_object.int_prop, $expected);
                    }
                }
            )*
        };
    }

    binop_test!(
        Int, addition: ("prop + 1", 6),
        Int, addition2: ("1 + prop", 6),
        Int, multiplication: ("prop * 2", 10),
        Int, multiplication2: ("2 * prop", 10),
        Int, subtraction: ("prop - 1", 4),
        Int, subtraction2: ("1 - prop", -4),
        Int, division: ("prop // 3", 1),
        Int, division2: ("15 // prop", 3),
        Int, modulo: ("15 %% prop", 0),
        Int, modulo2: ("16 %% prop", 1),
        Int, modulo3: ("prop %% 2", 1),
        Int, modulo4: ("prop %% (0 - 2)", 1),

        Bool, eqeq: ("prop == true", 1),
        Bool, eqeq2: ("prop == false", 0),
        Bool, eqeq3: ("prop == prop", 1),
        Bool, eqeq4: ("false == prop", 0),
        Bool, eqeq5: ("true == prop", 1),
        Bool, eqeq6: ("5 == 5", 1),

        Bool, neeq: ("true != true", 0),
        Bool, neeq2: ("true != false", 1),

        Bool, gt: ("10 > 10", 0),
        Bool, gt2: ("5 > 10", 0),
        Bool, gt3: ("15 > 10", 1),

        Bool, gteq: ("10 >= 10", 1),
        Bool, gteq2: ("5 >= 10", 0),
        Bool, gteq3: ("15 >= 10", 1),

        Bool, lt: ("10 < 10", 0),
        Bool, lt2: ("5 < 10", 1),
        Bool, lt3: ("15 < 10", 0),

        Bool, lteq: ("10 <= 10", 1),
        Bool, lteq2: ("5 <= 10", 1),
        Bool, lteq3: ("15 <= 10", 0),

        Int, then: ("5 then 8", 8),
        Int, then2: ("prop then 7", 7),
        Int, then3: ("7 then prop", 5),
    );

    #[derive(Serialize, Clone, Debug)]
    struct PropObj {
        int_prop: i32,
    }

    unsafe impl TapirScript for PropObj {
        fn set_prop(&mut self, index: u8, value: i32) {
            if index != 0 {
                panic!("Invalid index {index}");
            }

            self.int_prop = value;
        }

        fn get_prop(&self, index: u8) -> i32 {
            if index != 0 {
                panic!("Invalid index {index}");
            }

            self.int_prop
        }

        type EventType = ();

        fn script(self) -> Script<Self> {
            unimplemented!("Shouldn't create the script this way in the tests")
        }

        fn create_event(&self, _index: u8, _args: &[i32]) -> Self::EventType {}
    }
}
