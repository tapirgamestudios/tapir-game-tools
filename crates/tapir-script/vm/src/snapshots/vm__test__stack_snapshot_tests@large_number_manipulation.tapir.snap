---
source: crates/tapir-script/vm/src/lib.rs
expression: stack_at_waits
input_file: crates/tapir-script/vm/src/snapshot_tests/stack/large_number_manipulation.tapir
---
[
  ([
    [
      -1,
      6400,
    ],
  ], PropObj(
    int_prop: 3024,
  )),
  ([
    [
      -1,
      6400,
      33554440,
      6400,
    ],
  ], PropObj(
    int_prop: 3024,
  )),
  ([], PropObj(
    int_prop: 3024,
  )),
]
